torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1662,7 +1682,31 @@ struct to_ir {
         graph->createStore(x, fv)->insertBefore(false_block->return_node());
       }
 
-      auto unified = unifyTypes(tv->type(), fv->type());
+      // If we already know that the type is a Union, then we want to
+      // allow heterogeneous types to be unified
+      c10::optional<TypePtr> full_type = environment_stack->getType(x);
+      bool default_to_union =
+          full_type && (*full_type)->kind() == UnionType::Kind;
+      auto unified = unifyTypes(
+          tv->type(), fv->type(), /*default_to_union=*/default_to_union);
+
+      // Special logic for if we have a Union-typed value: check if
```

Comment:
This does not follow with the rest of our type checking. 
```
def foo(x: Optional[int]) -> int:
    if x is None:
         x = 4
    return x
```
It is important to be able to refine types. We should remove this logic.